x/mongo/driver/auth/oidc.go

```diff
@@ -0,0 +1,329 @@
+}
+
+func (*oidcOneStep) Completed() bool {
+	return true
+}
+
+func (oa *OIDCAuthenticator) providerCallback() (OIDCCallback, error) {
+	env, ok := oa.AuthMechanismProperties[environmentProp]
+	if !ok {
+		return nil, nil
+	}
+
+	switch env {
+	// TODO GODRIVER-2728: Automatic token acquisition for Azure Identity Provider
+	// TODO GODRIVER-2806: Automatic token acquisition for GCP Identity Provider
+	// This is here just to pass the linter, it will be fixed in one of the above tickets.
+	case azureEnvironmentValue, gcpEnvironmentValue:
+		return func(ctx context.Context, args *OIDCArgs) (*OIDCCredential, error) {
+			return nil, fmt.Errorf("automatic token acquisition for %q not implemented yet", env)
+		}, fmt.Errorf("automatic token acquisition for %q not implemented yet", env)
+	}
+
+	return nil, fmt.Errorf("%q %q not supported for MONGODB-OIDC", environmentProp, env)
+}
+
+// This should only be called with the Mutex held.
+func (oa *OIDCAuthenticator) getAccessToken(
+	ctx context.Context,
+	args *OIDCArgs,
+	callback OIDCCallback,
+) (string, error) {
+	if oa.accessToken != "" {
+		return oa.accessToken, nil
+	}
+
+	cred, err := callback(ctx, args)
+	if err != nil {
+		return "", err
+	}
+
+	oa.accessToken = cred.AccessToken
+	oa.tokenGenID++
+	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+	if cred.RefreshToken != nil {
+		oa.refreshToken = cred.RefreshToken
+	}
+	return cred.AccessToken, nil
+}
+
+// TODO GODRIVER-3246: Implement OIDC human flow
+// This should only be called with the Mutex held.
+//func (oa *OIDCAuthenticator) getAccessTokenWithRefresh(
+//	ctx context.Context,
+//	callback OIDCCallback,
+//	refreshToken string,
+//) (string, error) {
+//
+//	cred, err := callback(ctx, &OIDCArgs{
+//		Version:      apiVersion,
+//		IDPInfo:      oa.idpInfo,
+//		RefreshToken: &refreshToken,
+//	})
+//	if err != nil {
+//		return "", err
+//	}
+//
+//	oa.accessToken = cred.AccessToken
+//	oa.tokenGenID++
+//	oa.cfg.Connection.SetOIDCTokenGenID(oa.tokenGenID)
+//	return cred.AccessToken, nil
+//}
+
+// invalidateAccessToken invalidates the access token, if the force flag is set to true (which is
+// only on a Reauth call) or if the tokenGenID of the connection is greater than or equal to the
+// tokenGenID of the OIDCAuthenticator. It should never actually be greater than, but only equal,
+// but this is a safety check, since extra invalidation is only a performance impact, not a
+// correctness impact.
+// This must only be called with the lock held
+func (oa *OIDCAuthenticator) invalidateAccessToken() {
+	tokenGenID := oa.cfg.Connection.OIDCTokenGenID()
+	if tokenGenID >= oa.tokenGenID {
+		oa.accessToken = ""
+		oa.cfg.Connection.SetOIDCTokenGenID(0)
+	}
+}
+
+// Reauth reauthenticates the connection when the server returns a 391 code. Reauth is part of the
+// driver.Authenticator interface.
+func (oa *OIDCAuthenticator) Reauth(ctx context.Context) error {
```

Comment:
`Reauth` needs to take accept an `AuthConfig` that has all the inputs for authenticating a specific connection, not the connection most recently configured in `oa.cfg`.

E.g.
```go
func (oa *OIDCAuthenticator) Reauth(
	ctx context.Context,
	cfg *AuthConfig,
) error {
```

It's not clear exactly how to get all the info for an `AuthConfig` in `Operation.Execute`.

Reply:
I don't see how we can make that work. The python and rust driver both store the necessary information in the Authenticator :/